src/engine/source/builder/builders/opBuilderHelperMap.cpp

```diff
@@ -986,62 +986,184 @@ base::Expression opBuilderHelperRegexExtract(const std::string& targetField,
 //*           Array tranform                      *
 //*************************************************
 
-// field: +array_append/$field|literal...
-base::Expression opBuilderHelperAppend(const std::string& targetField,
-                                       const std::string& rawName,
-                                       const std::vector<std::string>& rawParameters,
-                                       std::shared_ptr<defs::IDefinitions> definitions)
+// field: array_append_unique($ref|literal)
+// field: array_append($ref|literal)
+HelperBuilder getBuilderArrayAppend(bool unique, std::shared_ptr<schemf::ISchema> schema)
 {
-    auto parameters = helper::base::processParameters(rawName, rawParameters, definitions);
+    return [=](const std::string& targetField,
+               const std::string& rawName,
+               const std::vector<std::string>& rawParameters,
+               std::shared_ptr<defs::IDefinitions> definitions)
+    {
+        auto parameters = helper::base::processParameters(rawName, rawParameters, definitions);
 
-    checkParametersMinSize(rawName, parameters, 1);
+        checkParametersMinSize(rawName, parameters, 1);
 
-    const auto name = helper::base::formatHelperName(rawName, targetField, parameters);
-
-    // Tracing
-    const std::string successTrace {fmt::format(TRACE_SUCCESS, name)};
+        const auto name = helper::base::formatHelperName(rawName, targetField, parameters);
 
-    const std::string failureTrace {fmt::format("[{}] -> Failure: ", name)};
-    const std::string failureSuffix1 {"Parameter '{}' reference not found"};
-    const std::string failureSuffix2 {"Parameter '{}' type is not a string"};
-    const std::string failureSuffix3 {"Parameter '{}' type unexpected"};
+        // Validate schema types and homogeneities
+        if (schema->hasField(targetField) && schema->getType(targetField) != json::Json::Type::Array)
+        {
+            throw std::runtime_error(
+                fmt::format("\"{}\" function: Target field \"{}\" is not an array", name, targetField));
+        }
 
-    // Return result
-    return base::Term<base::EngineOp>::create(
-        name,
-        [=, targetField = std::move(targetField)](base::Event event) -> base::result::Result<base::Event>
+        bool first = true;
```

Comment:
improve - desing: to follow this part is a few complex.
The logic could be reorganized so that reference parameters and regular parameters are treated in separate blocks to avoid repetitions.
You could use just one variable like type = json::Json::Type type; but initialized in json::Json::Type::Null for example. If type is still json::Json::Type::Null, we set it to the type of the first parameter found. If type has already been set, we compare the current type with type and throw an exception if they do not match.

These are suggestions that I think would make the code more readable, please let me know if any of this is not possible.

Reply:
The problem here is that null is a valid type, that's why we don't mix up the use of the variable. The type variable is used to identify the type (and all types are valid) and the first variable is used to know if the type is set.